bpf/lib/lb.h

```diff
@@ -935,6 +935,7 @@ static __always_inline int lb6_local(const void *map, struct __ctx_buff *ctx,
 			goto drop_no_service;
 		state->backend_id = backend_id;
 		ct_update_backend_id(map, tuple, state);
+		ct_update_rev_nat_index(map, tuple, svc->rev_nat_index);
```

Comment:
> Also, both ct_update_backend_id and ct_update_rev_nat_index seem to be used always together, so how about we optimize out one map_lookup_elem, combine these two functions into one and just pass the new updated state?

ack, that's a nice follow-on cleanup! Will queue up for the next PR.

Reply:
>I struggle to understand the motivation why you replaced state with raw rev_nat_index in the previous commit (this should be equivalent, because we assign state->rev_nat_index = svc->rev_nat_index a few lines below anyway), at the same time kept two similar snippets inconsistent. 

As you noted, the `state->rev_nat_index = svc->rev_nat_index` updates are mostly redundant - as we update it either way in the `update_state` epilogue. So I can obviously still add it along with the new `ct_update_rev_nat_index()` to keep things consistent, but didn't want to add unnecessary code. Any preference?

Reply:
>Also a little bit suspicious that we don't check !backend_id in this block, unlike the previous one. Is it redundant for some reason? I'm just thinking in the direction of further deduplicating these two blocks.

yeah that's why I'm slowly chipping away at the whole code path - to get a better picture of what other bugs are still lurking. Some of the `!backend_id` checks are just to respect session-affinity (which we don't want in this fallback path) - but I agree, that doesn't explain everything.

Reply:
>I think, you can remove the assignment under the update_state label, we jump there from CT_NEW where this assignment was already made, and from CT_REPLY where we also always end up with equal values (as long as you add it with the new ct_update_rev_nat_index).

Huh. The calling context expects to always find a valid `ct_state->rev_nat_index` - so I think removing the assignment in `update_state` will be slightly more involved. But sure, let's try.

Reply:
Note how the `CT_REPLY` case will typically *not* do the assignment (it checks for `state->rev_nat_index == 0`, not for eg. `state->rev_nat_index == svc->rev_nat_index`). So in its common path, it currently relies on the assignment at `update_state`.

Reply:
>ack, that's a nice follow-on cleanup! Will queue up for the next PR.

stashed in https://github.com/cilium/cilium/pull/25372 :heart: